File: CellLine.py
import random as rnd
from Cell import Cell

class CellLine:
    """Handles the specimens of a specific cell lineage.
    A cell lineage is a group of cells that have a common ancestor, we represent the common \
    ancestor by it's genome, so, in this setting, the cells in a cell line share a common \
    ancestral genome.
    This structure is in charge of holding this ancestral code and managing the cells \
    creating new cells when needed and cleaning up the dead ones.
    
    
    Atributes
    ---------    
            + Ancestral genome: A string to which all cells genomes come from.
            + Cells: The group cells belonging to this cell line.
            + Alive cells and Dead cells: The cells that are respectively currently alive or dead
            + System: The system this cells form a part of.
            + Current cell index: Each cell has a unique identification label (index). 
                                  This is the label to place in the next cell to be born
    """
    
    def __init__(self, 
                 system, 
                 genome = None, 
                 recycleDeadCells=True, 
                 genomeAlphabet=None):
        """ A cell line is instantiated with an obligatory system, \
        to which the cell line belongs. Optionals are the genome and the \
        genome alphabet. If a custom genome is passed, the genome alphabet \
        should be passed too unless it is formed of the letters in "ACGT".
        Also, dead cells are recycled by default, this helps to improve \
        memory usage. 
        """
        self.recycleDeadCells = recycleDeadCells
        self.currentIndex = 0
        self.cells = []
        self.aliveCells = set()
        self.deadCells = set()
        self.system = system
        # Set the default genome
        if not genome:
            genome = 10*'A'
        self.genome = genome
        # Check if the genome alphabet option was set
        if not genomeAlphabet:
            genomeAlphabet = "ACGT"
        self.genomeAlphabet = genomeAlphabet
        # Validate the resulting genome/genome alphabet combination
        for letter in genome:
            if letter not in genomeAlphabet:
                raise ValueError('`genome` must contain only letters present in the `genomeAlphabet`')
    # ---
        
    def addCellTo(self, site):
        """Get a new, initialized cell
        """
        # Create the new cell
        new = self.newCell()
        new.addTo(site)
        
        return new # Return the new cell
    # ---
        
    def newCell(self):
        """ Get a new blank cell in this lineage and system.
        """
        # Fetch a blank cell
        if self.recycleDeadCells and self.deadCells:
            # Fetch a dead cell to recycle
            new = self.fetchCellToRecycle()
            self.recycleCell(new) 
        else:
            new = Cell(system = self.system, 
                       cellLine = self, 
                       index = self.currentIndex)
            self.currentIndex += 1
            self.cells.append( new )
            
        # Update state to take new cell into account
        self.aliveCells.add( new )
        return new
    # ---
    
    def fetchCellToRecycle(self):
        recycled = self.deadCells.pop()
        return recycled
    # ---
    
    def recycleCell(self, cell):
        cell.setIndex(self.currentIndex)
        self.currentIndex += 1
        
    
    def getAliveCells(self):
        return self.aliveCells
    # ---
    
    def totalCells(self):
        return len(self.cells)
    # ---
    
    def totalAliveCells(self):
        return len(self.aliveCells)
    # ---
    
    def totalDeadCells(self):
        return len(self.deadCells)
    # ---
    
    def sampleCells(self, all=False, n=1):
        # If it is asked for all cells, 
        # take a sample of the whole size
        if all==True:
            n = len(self.aliveCells)
        # If the default argument is set, return a single cell
        elif n == 1:
            return rnd.sample(self.aliveCells, n) [0]
        # Return a sample of size n
        return rnd.sample(self.aliveCells, n)    
    # ---
        
    def handleDeath(self, dying):
        """ Process the dying cell. This means removing from the alive cells\
        and adding to the dead ones, maybe to recicle it when another is born.
        """
        self.aliveCells.remove(dying)
        self.deadCells.add(dying)
        # Cell is now officially dead
    # ---
    
    def getGenomeAlphabet(self):
        return self.genomeAlphabet
    # ---
    
    def getBaseGenome(self):
        return self.genome
    # ---
    
    def getGenomeLength(self):
        return len(self.genome)
    

File: c/src/simulation/simplecells/cell_line.c
 
/**
 * =========
 * Cell line
 * =========
 * 
 * Implementation file for the CellLine class, 
 * along with methods and auxiliary definitions.
 */

#include <stdlib.h>
#include <stdio.h>

#include "cell_line.h"
#include "cell/cell.h"



// ----- Main structure

// The structure is private and should not be accessed directly 
// from outside this file, but with the interface described in the 
// header file.  
struct CellLine_struct { /**
    Handles the specimens of a specific cell lineage.

    A cell lineage is a group of cells that have a common ancestor, we
    represent the common ancestor by it's ancestral genome.
    This structure is in charge of holding this ancestral code and managing
    the cells creating new cells when needed and cleaning up the dead ones.

    Atributes:
            + Ancestral genome: A string-like object.
            + Cells: The cells inherited from this cell line.
            + Current cell index: Each cell has a unique index. This is the
                                  index to place in the next cell to be born.
    */
    int current_index;
    List *cells;
};

typedef struct CellLine_struct CellLine;



// ----- Methods

CellLine *CellLine_new( ) { /**
    Creation of a cell lineage.
    */
    // Create
    CellLine *self = malloc(1 * sizeof(*self));
    // Initialize
    self->current_index = 0;
    self->cells = List_new();
        
    return self;
        
} // --- CellLine_new


void CellLine_del( CellLine *self ) { /**
    Deleter for a cell line object.
    */
    // 1 --- Delete the cells from this cell line
        List *cells = self->cells;
        
        for(int i=0; i < List_size(cells); i++) {
            // Fetch the ith cell
            Cell *item = List_at(cells, i);
            // Delete the cell
            Cell_del(item);
        }
        
        // Remove the list
        List_del(cells);
        
    // 2 --- Delete the cell line structure
        free(self);
        
} // --- CellLine_del


Cell *CellLine_new_cell( CellLine *self) { /**
    Get a new blank cell in this lineage and system.
    */
    // Fetch a fresh, new cell
    Cell *cell = Cell_new( self,              // Cell line
                           self->current_index ); // Index
        
    self->current_index += 1;
        
    // Add to the cells from this cell line
    List_push(self->cells, cell);
        
    return cell;
    
} // --- CellLine_new_cell


Cell *CellLine_recycle_cell(CellLine *self, 
                            Cell *cell, 
                            int father) { /**
    Clear previous information from a cell.
    */
    Cell_set_father(cell, father);
        
    Cell_set_index(cell, self->current_index);
        
    self->current_index += 1;
        
    return cell;
        
} // --- CellLine_recycle_cell


CellLine *CellLine_process( CellLine *self ) { /**
    Move a step forward in time.
    */
    List *cells = self->cells;
    int n = List_size(cells);
    
    for( int i=0; i<n; i++ ) {
        // Fetch the i_th cell
        Cell *cell = List_at(cells, i);
        // Process the cell
        Cell_process(cell);
    }
    
    return self;
            
} // --- CellLine_process


void CellLine_print( CellLine *self ) { /**
    Print the state of the cell line.
    */
    List *cells = self->cells;
    int n = List_size(cells);
    
    printf("<CellLine object. Cells: [");
    
    for( int i=0; i<n; i++ ) {
        // Fetch the i_th cell
        Cell *cell = List_at(cells, i);
        // Process the cell
        printf("%d, ", Cell_index(cell));
    }
    
    printf("]>");
    
} // --- CellLine_print
